src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,290 @@
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "abstract_operator.hpp"
+#include "all_type_variant.hpp"
+#include "resolve_type.hpp"
+#include "storage/dictionary_segment.hpp"
+#include "storage/reference_segment.hpp"
+#include "storage/value_segment.hpp"
+#include "table_scan.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value)
+    : _in{in}, _column_id{column_id}, _scan_type{scan_type}, _search_value{search_value} {}
+
+ColumnID TableScan::column_id() const { return _column_id; }
+
+ScanType TableScan::scan_type() const { return _scan_type; }
+
+const AllTypeVariant& TableScan::search_value() const { return _search_value; }
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  auto in_table_ptr = _in->get_output();
+  auto n_chunks = in_table_ptr->chunk_count();
+  auto data_type = in_table_ptr->column_type(_column_id);
+
+  // accumulate the scan results here. We will have one chunk
+  // for each chunk in the input table (that has at least one
+  // row matching the filter).
+  auto result_chunks_ptr = std::make_shared<std::vector<std::shared_ptr<Chunk>>>();
+
+  // iterate over all chunks in the input table and process each chunk seperately.
+  // We first determine which rows should be included in the output table,
+  // i.e. which rows match the filter condition. Then, we construct new
+  // chunks that consist of reference segments and make up the output table.
+  for (auto chunk_id = ChunkID{0}; chunk_id < n_chunks; ++chunk_id) {
+    auto chunk_ptr = in_table_ptr->get_chunk(chunk_id);
+    auto include_rows_ptr = scan_chunk(chunk_ptr, chunk_id, data_type);
```

Comment:
You basically build your PosLists multiple times: `scan_chunk` returns a vector of `ChunkOffset`s that you iterate again in `subset_chunk` to build (multiple) PosLists. Instead, you could build one single PosList in `scan_chunk`.